crates/amaru-ledger/src/state/transaction.rs

```diff
@@ -204,78 +207,103 @@ impl InputsOutputs<true> for MintedTransactionBody<'_> {
+fn flatten_certificate(certificate: Certificate) -> Vec<Certificate> {
+    match certificate {
+        Certificate::StakeVoteDeleg(stake_credential, hash, drep) => {
+            vec![
+                Certificate::VoteDeleg(stake_credential.clone(), drep),
+                Certificate::StakeDelegation(stake_credential, hash),
+            ]
+        }
+        Certificate::StakeRegDeleg(stake_credential, hash, coin) => {
+            vec![
+                Certificate::Reg(stake_credential.clone(), coin),
+                Certificate::StakeDelegation(stake_credential, hash),
+            ]
+        }
+        Certificate::StakeVoteRegDeleg(stake_credential, hash, drep, coin) => {
+            vec![
+                Certificate::Reg(stake_credential.clone(), coin),
+                Certificate::StakeDelegation(stake_credential.clone(), hash),
+                Certificate::VoteDeleg(stake_credential, drep),
+            ]
+        }
+        Certificate::VoteRegDeleg(stake_credential, drep, coin) => {
+            vec![
+                Certificate::Reg(stake_credential.clone(), coin),
+                Certificate::VoteDeleg(stake_credential, drep),
+            ]
+        }
+        _ => vec![certificate],
+    }
+}
+
 fn apply_certificates(
     parent: &Span,
     pools: &mut DiffEpochReg<PoolId, PoolParams>,
     accounts: &mut DiffBind<StakeCredential, PoolId, Lovelace>,
     certificates: Vec<Certificate>,
 ) {
-    for certificate in certificates {
+    certificates
+        .into_iter()
+        .flat_map(flatten_certificate)
+        .for_each(|certificate| {
         match certificate {
-                Certificate::StakeRegistration(credential) | Certificate::Reg(credential, ..) | Certificate::VoteRegDeleg(credential, ..) => {
-                    trace!(name: "certificate.stake.registration", target: EVENT_TARGET, parent: parent, credential = ?credential);
-                    accounts.register(credential, STAKE_CREDENTIAL_DEPOSIT as Lovelace, None);
-                }
-                Certificate::StakeDelegation(credential, pool)
-                // FIXME: register DRep delegation
-                | Certificate::StakeVoteDeleg(credential, pool, ..) => {
-                    trace!(name: "certificate.stake.delegation", target: EVENT_TARGET, parent: parent, credential = ?credential, pool = %pool);
-                    accounts.bind(credential, Some(pool));
-                }
-                Certificate::StakeRegDeleg(credential, pool, ..)
-                // FIXME: register DRep delegation
-                | Certificate::StakeVoteRegDeleg(credential, pool, ..) => {
-                    trace!(name: "certificate.stake.registration", target: EVENT_TARGET, parent: parent, credential = ?credential);
-                    trace!(name: "certificate.stake.delegation", target: EVENT_TARGET, parent: parent, credential = ?credential, pool = %pool);
-                    accounts.register(
```

Comment:
@KtorZ this is now split into 2 operations (register then bind) so that the last argument of `register` is always `None`. Unclear about the implication: do we want to remove this param and simplify `DiffBind` ?

Looks to me that this simplification is the right direction, even at the cost of removing the opportunity for those kind of optimizations.